lambda-buffers-codegen/src/LambdaBuffers/Codegen/Rust/Print/LamVal.hs

```diff
@@ -0,0 +1,397 @@
+printProductE iTyDefs (qtyN@(mn', tyN'), _) vals = do
+  let mn = withInfo mn'
+      tyN = withInfo tyN'
+
+  let ctorDoc = R.printMkCtor tyN
+  (tyArgs, fieldTys) <-
+    case Map.lookup qtyN iTyDefs of
+      Just (PC.TyDef _ (PC.TyAbs tyArgs (PC.ProductI (PC.Product fields _)) _) _) -> return (toList tyArgs, fields)
+      _ -> throwInternalError "Expected a ProductE but got something else (TODO(szg251): Print got)"
+  let phantomFieldDocs = R.printRsQTyName RR.phantomData <$ TD.collectPhantomTyArgs fieldTys tyArgs
+      mayBoxedFields = zip vals $ TD.isRecursive iTyDefs mn tyN <$> fieldTys
+
+  fieldDocs <- for mayBoxedFields (printMaybeBoxed iTyDefs)
+
+  return $ ctorDoc <> encloseSep lparen rparen comma (fieldDocs <> phantomFieldDocs)
+
+printMaybeBoxed :: MonadPrint m => PC.TyDefs -> (LV.ValueE, Bool) -> m (Doc ann)
+printMaybeBoxed iTyDefs (val, False) = clone <$> printValueE iTyDefs val
+printMaybeBoxed iTyDefs (val, True) = do
+  valDoc <- clone <$> printValueE iTyDefs val
+  return $ R.printRsQValName RR.boxNew <> parens valDoc
+
+printTupleE :: MonadPrint m => PC.TyDefs -> LV.ValueE -> LV.ValueE -> m (Doc ann)
+printTupleE iTyDefs l r = do
+  lDoc <- printValueE iTyDefs l
+  rDoc <- printValueE iTyDefs r
+  return $ parens (lDoc <> comma <+> rDoc)
+
+printTextE :: MonadPrint m => Text.Text -> m (Doc ann)
+printTextE = return . fromStr . dquotes . pretty
+
+printCaseTextE :: MonadPrint m => PC.TyDefs -> LV.ValueE -> [(LV.ValueE, LV.ValueE)] -> (LV.ValueE -> LV.ValueE) -> m (Doc ann)
+printCaseTextE iTyDefs txtVal cases otherCase = do
+  caseValDoc <- printValueE iTyDefs txtVal
+  caseDocs <-
+    for
+      cases
+      ( \case
+          (LV.TextE caseTxt, bodyVal) -> do
+            conditionDoc <- printTextE caseTxt
+            bodyDoc <- printValueE iTyDefs bodyVal
+            return $ group $ conditionDoc <+> "=>" <+> bodyDoc
+          (_wrongCaseVal, _) -> throwInternalError "Expected a TextE as the case value but got something else (TODO(bladyjoker): Print got)"
+      )
+  otherDoc <- printOtherCase iTyDefs otherCase
+  return $ "ma" <> align ("tch" <+> caseValDoc <+> braces (line <> vsep (punctuate comma (caseDocs <> [otherDoc]))))
+
+{- | Print a reference
+
+ To help Rust type inference, we inject type information for a few references.
+ In case of a `toPlutusData`, `fromPlutusData`, `toJson`, `fromJson` we call the reference as a trait
+ method on the target type:
+ ```rs
+ <TargetType>::to_plutus_data(...)
+ ```
+ In case of `jsonConstructor`, we print the target type in a turbofish syntax:
+ ```rs
+ lbr_prelude::json::json_constructr::<TargetType>::toPlutusData(...)
+ ```
+-}
+printRefE :: MonadPrint m => LV.Ref -> m (Doc ann)
+printRefE ref = do
+  qvn <- LV.resolveRef ref
+  case ref of
+    (argTy : _, builtin)
+      | builtin == "toPlutusData"
+          || builtin == "fromPlutusData"
+          || builtin == "toJson"
+          || builtin == "fromJson" -> do
```

Comment:
why not just have a standalone function

```
toPlutusData :: IsPlutusData a => a -> PlutusData
toPlutusData x = x.toPlutusData()
```

And then print `toPlutusData x`?

Imagine if that was Rust.

`<Type>::<method> is a convenience syntax afaik?`

Reply:
I'll try to solve these hacks later!